core/src/test/java/com/linecorp/armeria/server/ServerMaxConnectionAgeTest.java

```diff
@@ -0,0 +1,247 @@
+            sb.requestTimeoutMillis(0);
+            sb.maxConnectionAgeMillis(MAX_CONNECTION_AGE);
+            sb.service("/", (ctx, req) -> HttpResponse.of(OK));
+            sb.service("/slow", (ctx, req) ->
+                    HttpResponse.delayed(HttpResponse.of(OK), Duration.ofMillis(400)));
+        }
+
+        @Override
+        protected boolean runForEachTest() {
+            return true;
+        }
+    };
+
+    @RegisterExtension
+    static ServerExtension serverKeepAlive = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.idleTimeoutMillis(0);
+            sb.requestTimeoutMillis(0);
+            sb.service("/", (ctx, req) ->
+                    HttpResponse.delayed(HttpResponse.of(OK), Duration.ofMillis(100)));
+        }
+    };
+
+    private List<HttpResponse> responses;
+    private AtomicInteger opened;
+    private AtomicInteger closed;
+    private ConnectionPoolListener connectionPoolListener;
+
+    @BeforeEach
+    void setUp() {
+        responses = new ArrayList<>();
+        opened = new AtomicInteger();
+        closed = new AtomicInteger();
+        connectionPoolListener = new ConnectionPoolListener() {
+            @Override
+            public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                       InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                opened.incrementAndGet();
+            }
+
+            @Override
+            public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                         InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                closed.incrementAndGet();
+            }
+        };
+    }
+
+    @Test
+    void http1MaxConnectionAge() throws InterruptedException {
+        final AtomicInteger oldClosed = new AtomicInteger();
+        final WebClient client = newWebClient(server.uri(SessionProtocol.H1C));
+
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        while (closed.get() < 5) {
+            final AggregatedHttpResponse agg = client.get("/").aggregate().join();
+            assertThat(agg.status()).isEqualTo(OK);
+
+            final int closed = this.closed.get();
+            assertThat(opened).hasValueBetween(closed, closed + 1);
+
+            // When a connection is disconnected, check the time from the previous disconnection.
+            if (this.closed.get() > oldClosed.get()) {
+                assertThat(stopwatch.elapsed().toMillis())
+                        .isCloseTo(MAX_CONNECTION_AGE, withinPercentage(35));
+                oldClosed.set(this.closed.get());
+                stopwatch.reset().start();
+            }
+        }
+    }
+
+    // TODO(ikhoon): This is a flaky test. Needs a idea to fix it.
+    @Test
+    void http1WithPipeliningMaxConnectionAge() {
```

Comment:
How about this:

- Send the first streaming POST response with a few bytes written.
  - The response stream must be stay open.
- Wait until the first response's first few bytes were drained by Armeria.
  - At this point, we are sure the connection has been returned to the pool.
- Send the second non-streaming GET response.
- Wait until the connection reaches its max age.

Reply:
To test HTTP/1.1 pipeline, I used Socket. Unfortunately, it was hard to control exact request timing with Armeria WebClient. 😅

Reply:
:+1: